kafka/integration_test.go

```diff
@@ -0,0 +1,529 @@
+		return
+	}
+	low, high, err := p.QueryWatermarkOffsets(testconf.Topic, 0, 5*1000)
+	if err != nil {
+		t.Errorf("Failed to query watermark offsets for topic %s\n", testconf.Topic)
+		return
+	} else {
+		t.Logf("Watermark offsets fo topic %s: low=%d, high=%d\n", testconf.Topic, low, high)
+	}
+}
+
+// TestProducerFuncBasedAPI test function-based API, requires broker
+func TestProducerFuncBasedAPI(t *testing.T) {
+	if !testconfRead() {
+		t.Skipf("Missing testconf.json")
+	}
+
+	kConfig := &ConfigMap{
+		"bootstrap.servers":    testconf.Brokers,
+		"default.topic.config": ConfigMap{"produce.offset.report": true},
+	}
+
+	p, err := NewProducer(kConfig)
+	if err != nil {
+		t.Fatalf("Failed to create producer: %s\n", err)
+	}
+
+	t.Logf("Created Producer %v\n", p)
+
+	deliveryChan := make(chan Event)
+
+	for _, testmsg := range testmsgs {
+		m := &Message{
+			TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: testmsg.Partition},
+			Value:          testmsg.Value,
+			Key:            testmsg.Key}
+
+		// coverage test for message String() API
+		msgContent := m.String()
+		t.Logf("Producing message %s\n", msgContent)
+
+		err = p.Produce(m, deliveryChan)
+		if err != nil {
+			if testmsg.ExpectedError != "" && err.Error() == testmsg.ExpectedError {
+				fmt.Fprintf(os.Stderr, "OK: Expected error: %s\n", testmsg.ExpectedError)
+				continue
+			} else {
+				t.Errorf("Encountered error: %s\n", err)
+				continue
+			}
+		}
+
+		e := <-deliveryChan
+		verifyEvent(e, testmsg, t)
+	}
+	p.Close()
+}
+
+// TestProducerChannelBasedAPI test channel-based API, requires broker
+func TestProducerChannelBasedAPI(t *testing.T) {
+	runProducerChannelBasedAPI(t)
+}
+
+// run the prodcuer using channel based API
+func runProducerChannelBasedAPI(t *testing.T) {
+	if !testconfRead() {
+		t.Skipf("Missing testconf.json")
+	}
+
+	kConfig := &ConfigMap{
+		"bootstrap.servers":    testconf.Brokers,
+		"default.topic.config": ConfigMap{"produce.offset.report": true}}
+
+	p, err := NewProducer(kConfig)
+	if err != nil {
+		t.Fatalf("Failed to create producer: %s\n", err)
+	}
+
+	t.Logf("Created Producer %v\n", p)
+
+	var doneChan chan bool
+
+	for _, testmsg := range testmsgs {
+
+		doneChan = make(chan bool)
+
+		// delivery handler
+		go func() {
+			e := <-p.Events()
+			verifyEvent(e, testmsg, t)
+			close(doneChan)
+		}()
+
+		m := &Message{
+			TopicPartition: TopicPartition{Topic: &testconf.Topic, Partition: testmsg.Partition},
+			Value:          testmsg.Value,
+			Key:            testmsg.Key}
+		p.ProduceChannel() <- m
+
+		// wait for delivery report goroutine to finish
+		_ = <-doneChan
+
+	}
+
+	p.Close()
+
+}
+
+// verify event info
+func verifyEvent(e Event, testmsg testmsgType, t *testing.T) {
+	switch e.(type) {
+	case *Message:
+		m := e.(*Message)
+		if m.TopicPartition.Error != nil {
+			err := m.TopicPartition.Error
+			if testmsg.ExpectedError != "" && err.Error() == testmsg.ExpectedError {
```

Comment:
I think it might be better to have the ExpectedError as a Kafka Error and matching directly on that, rather than the string representation.
I believe that's more correct, showcases better API usage (for users that checkout the tests for inspiration) and also makes the code cleaner.